### ISSUE TITLE
[Feature expansion] nestable HTML attributes

### ISSUE THREAD

**Issue by Faryshta** [yiisoft/yii2#7350]
I don't know if thats the proper term to use, but its what best I imagined to explain it.

Currently the `BaseHtml` can produce html tag attributes using 'options' and there is an special option `data` that is the only one allowed to contain options itself. Example:

```php
'options' => [
    'id' => 'modalButton',
    'class' => 'btn btn-default',
    'data' => [
        'toggle' => 'modal'
        'target' => '#loginModal',
    ],
]
```

Currently 'data' is the only attribute that can achieve this because its hard coded here https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseHtml.php#L1624

My feature expansion is to have an static property `BaseHtml::nestableAttributes = ['data']` which can be edited when extending the `BaseHtml` class. For example:

```php
AngularHtml extends BaseHtml
{
     $nestableAttributes = ['data', 'ng'];
}
```
Will allow me to do

```php
AngularHtml::textInput('search', null, [
     'ng' => [
          'model' => 'search',
          'pattern' => '/^\d+$/',
          'click' => 'searchClick()',
     ]
])
```

**Comment by samdark** [comment]
Is that for Angular? Overall it makes sense to allow customizing it.

**Comment by Faryshta** [comment]
Yup I made my example with angularjs in mind. Let me prepare a pr, i think it will just be a couple of lines.

Also, do you think an static method `BaseHtml::addNestableAttribute` could be useful?

**Comment by samdark** [comment]
No.

**Comment by mdmunir** [comment]
Related with this issue. If we work with easyui, we can add option via `data-options` attribute
```html
<input class="easyui-textbox" data-options="{required:true,validType:'email'}">
```
For now, we must send `$options` parameter as
```php
$options = [
    'class'=>'easyui-textbox',
    'data'=>[
        'options'=>['required'=>true,'validType'=>'email'],
    ]
];
Html::tag('input','',$options);
```
Much better if we can do like this
```php
$options = [
    'class'=>'easyui-textbox',
    'data-options'=>['required'=>true,'validType'=>'email'],
];
```

**Comment by qiangxue** [comment]
@mdmunir The current implementation should already allow you to do the 2nd version.

**Comment by Faryshta** [comment]
@qiangxue I think @mdmunir tries to use a shortcut. The current way to do it is: 

```php
'options' => [
    'data'=>[ 'options'=>['required'=>true,'validType'=>'email']]
]
```

And it will produce `data-options="{required:true,validType:'email'}"`

But if you try this shortcut:

```php
'options' => [
    'data-options'=>['required'=>true,'validType'=>'email']
]
```
You will not get any attribute at all.

I have an issue open for shortcuts in configuration #7303 but since this shortcut is a new functionality I don't think it will get implemented before 2.1

**Comment by mdmunir** [comment]
When we use 
```php
$options = [
    'class'=>'easyui-textbox',
    'data-options'=>['required'=>true,'validType'=>'email'],
];
```
it will produce error
```
PHP Warning – yii\base\ErrorException 
htmlspecialchars() expects parameter 1 to be string, array given
```
https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseHtml.php#L1624-L1632 should be
```php
            elseif (is_array($value) && $name === 'data') {
                foreach ($value as $n => $v) {
                    if (is_array($v)) {
                        $html .= " $name-$n='" . Json::encode($v, JSON_HEX_APOS) . "'";
                    } else {
                        $html .= " $name-$n=\"" . static::encode($v) . '"';
                    }
                }
            } elseif (is_array($value) && strpos($name, 'data-')===0) {
                if (is_array($value)) {
                    $html .= " $name='" . Json::encode($value, JSON_HEX_APOS) . "'";
                } else {
                    $html .= " $name=\"" . static::encode($value) . '"';
                }                
            } 
```
